### DISCUSSION TITLE
chisel-example using mill?

### DISCUSSION THREAD

**Discussion by kodakrome** [chipsalliance/chisel#4081] [Q&A]
I am able to do the quick-start and build/run `chisel-example.scala` using `scala-cli`, but many of the other projects available use sbt or mill.  The chisel documentaion recommends mill, but references sbt in a few places.

Can anyone help me with a basic `build.sc` for mill that makes `chisel-example.scala` work?  I think that might help me bridge the gap from `scala-cli` to building a more substantial design.

Thanks!

**Comment by jackkoenig** [comment]
The [chisel-template](https://github.com/chipsalliance/chisel-template) has Mill (with basic instructions). I'd like to add some instructions to the website but never seem to find the time 😅. I've done some experimentation with `g8` templates which would make it possible to create a project from the chisel-template on the command-line: https://github.com/jackkoenig/chisel-template.g8 (note that fork is out-of-date with upstream).

**Reply by kodakrome** [reply]
Thank you.  I have to admit `extends SbtModule` kinda threw me for a loop.
I'll try to understand what each of these lines does, but I think you are telling me that this setup should work for 90%+ of projects or designs I try to pull in.

**Reply by jackkoenig** [reply]
> I have to admit extends SbtModule kinda threw me for a loop.

`SbtModule` would be more precisely named `SbtFilesystemLayoutModule`--it just means that source files are in SBT's default `src/main/scala`[^1] rather than Mill's default `src`. Since the template supports both build tools it needs agreement on where the source files are.

> I'll try to understand what each of these lines does, but I think you are telling me that this setup should work for 90%+ of projects or designs I try to pull in.

It's sufficient setup for most projects. It doesn't handle source dependencies but that's a complex topic probably best handled by [Mill's own documentation](https://mill-build.com/mill/Intro_to_Mill.html) and also heavily dependent on the specific upstream project one would want to depend on.

[^1]: Derived from Maven's default `src/main/java`.
